### ISSUE TITLE
Simple use of KJ_REQUIRE causes SIGSEGV

### ISSUE THREAD

**Issue by jrsala-auguration** [capnproto/capnproto#1416]
The following code causes a segfault

```
#include <kj/debug.h>

int main() {
  const char* myPtr = nullptr;
  KJ_REQUIRE(myPtr != nullptr, "Some message");
}
```

`KJ_REQUIRE` constructs a `DebugComparison` containing a `const char*&` to the null `const char*` in its `left` data member, and then passes that to the `Debug::Fault` constructor which attempts to stringify it. This in turn ends up passing that null `const char*` to `kj::toCharSequence` which calls `strlen` on it.

I fixed this issue in my code by instead using `KJ_REQUIRE((myPtr != nullptr), "Some message");` but I suspect either stringifying a `const char*` should not so casually call `strlen`, or the overload of `KJ_STRINGIFY` for `DebugComparison` and `DebugExpression` arguments should be a bit more particular when it comes to stringifying their data members.

**Comment by kentonv** [comment]
I suppose a null `const char*` should probably stringify to "(null)" or something.

**Comment by jclee** [comment]
Good catch... Might be worth noting that not-null assertions in kj code are typically done by wrapping expressions in `KJ_REQUIRE_NONNULL(expr, "some message")`, which may explain why the issue hadn't become apparent sooner.

**Comment by kentonv** [comment]
Eh... `KJ_REQUIRE_NONNULL` is meant to be used with `kj::Maybe`, not pointers... if it works with pointers I think that's by accident.

**Comment by jclee** [comment]
Good to know.  Yeah, I was curious if it would work for pointers, so I dug a little into the implementation, and it looks like it should, although I guess that functionality was added mainly to support pointers in `KJ_IF_MAYBE()` rather than `KJ_REQUIRE_NONNULL()`.

https://github.com/capnproto/capnproto/blob/09b91a3/c%2B%2B/src/kj/common.h#L1266-L1268

**Comment by kentonv** [comment]
Yeah... and TBH the ability to use KJ_IF_MAYBE on pointers was something I added very early on and then, I think, never ever used anywhere, to the point that I forgot it was possible. Probably a misfeature.
